// @salesforce/packaging/lib/package1/packageVersion1GP.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersion1GP = void 0;
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const kit_1 = require("@salesforce/kit");
/**
 * Package1Version class - Class to be used with 1st generation package versions
 */
class PackageVersion1GP extends kit_1.AsyncCreatable {
    // @ts-ignore
    constructor(options) {
        super(options);
        this.options = options;
    }
    convert() {
        return Promise.resolve(undefined);
    }
    create() {
        return Promise.resolve(undefined);
    }
    delete() {
        return Promise.resolve(undefined);
    }
    install() {
        return Promise.resolve(undefined);
    }
    list() {
        return Promise.resolve(undefined);
    }
    uninstall() {
        return Promise.resolve(undefined);
    }
    update() {
        return Promise.resolve(undefined);
    }
    init() {
        return Promise.resolve(undefined);
    }
}
exports.PackageVersion1GP = PackageVersion1GP;
