// @salesforce/plugin-packaging/lib/commands/force/package/beta/version/create/list.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionCreateListCommand = void 0;
const os = require("os");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_version_create_list');
const packaging = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'packaging');
class PackageVersionCreateListCommand extends command_1.SfdxCommand {
    // eslint-disable-next-line @typescript-eslint/require-await
    async run() {
        throw new Error('Beta command not yet implemented');
    }
}
exports.PackageVersionCreateListCommand = PackageVersionCreateListCommand;
PackageVersionCreateListCommand.description = messages.getMessage('cliDescription');
PackageVersionCreateListCommand.examples = messages.getMessage('examples').split(os.EOL);
PackageVersionCreateListCommand.requiresDevhubUsername = true;
PackageVersionCreateListCommand.flagsConfig = {
    createdlastdays: command_1.flags.number({
        char: 'c',
        description: packaging.getMessage('createdLastDaysDescription'),
        longDescription: packaging.getMessage('createdLastDaysLongDescription'),
    }),
    status: command_1.flags.enum({
        char: 's',
        description: messages.getMessage('statusDescription'),
        longDescription: messages.getMessage('statusLongDescription'),
        options: ['Queued', 'InProgress', 'Success', 'Error'],
    }),
};
