// @salesforce/sf-plugins-core/lib/ux/ux.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ux = void 0;
const core_1 = require("@oclif/core");
const base_1 = require("./base");
class Ux extends base_1.UxBase {
    constructor(outputEnabled) {
        super(outputEnabled);
    }
    table(data, columns, options) {
        this.maybeNoop(() => core_1.CliUx.ux.table(data, columns, options));
    }
    url(text, uri, params = {}) {
        this.maybeNoop(() => core_1.CliUx.ux.url(text, uri, params));
    }
    styledJSON(obj) {
        this.maybeNoop(() => core_1.CliUx.ux.styledJSON(obj));
    }
    styledObject(obj) {
        this.maybeNoop(() => core_1.CliUx.ux.styledObject(obj));
    }
    styledHeader(text) {
        this.maybeNoop(() => core_1.CliUx.ux.styledHeader(text));
    }
}
exports.Ux = Ux;
