// @sf/functions/lib/commands/env/logdrain/remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const color_1 = require("@heroku-cli/color");
const core_1 = require("@oclif/core");
const cli_ux_1 = require("cli-ux");
const core_2 = require("@salesforce/core");
const flags_1 = require("../../../lib/flags");
const base_1 = require("../../../lib/base");
const utils_1 = require("../../../lib/utils");
core_2.Messages.importMessagesDirectory(__dirname);
const messages = core_2.Messages.loadMessages('@salesforce/plugin-functions', 'env.logdrain.remove');
class LogDrainRemove extends base_1.default {
    async run() {
        var _a, _b, _c, _d, _e, _f;
        const { flags } = await this.parse(LogDrainRemove);
        this.postParseHook(flags);
        // We support both versions of the flag here for the sake of backward compat
        const targetCompute = (_a = flags['target-compute']) !== null && _a !== void 0 ? _a : flags.environment;
        const url = (_b = flags['drain-url']) !== null && _b !== void 0 ? _b : flags.url;
        if (!targetCompute) {
            throw new core_1.Errors.CLIError(`Missing required flag:
        -e, --target-compute TARGET-COMPUTE  ${color_1.default.dim('Environment name.')}
       See more help with --help`);
        }
        if (!url) {
            this.error(new Error('Missing required flag: -l, --drain-url Logdrain url to remove'));
        }
        if (flags.environment) {
            this.warn(messages.getMessage('flags.environment.deprecation'));
        }
        if (flags.url) {
            this.warn(messages.getMessage('flags.url.deprecation'));
        }
        const appName = await (0, utils_1.resolveAppNameForEnvironment)(targetCompute);
        try {
            cli_ux_1.cli.action.start(`Deleting drain for environment ${color_1.default.app(targetCompute)}`);
            await this.client.delete(`/apps/${appName}/log-drains/${encodeURIComponent(url)}`);
            cli_ux_1.cli.action.stop();
            return 'Removed drain-url';
        }
        catch (e) {
            const error = e;
            if ((_d = (_c = error.data) === null || _c === void 0 ? void 0 : _c.message) === null || _d === void 0 ? void 0 : _d.includes('Url is invalid')) {
                this.error(new Error(`URL is invalid ${url}`));
            }
            if ((_f = (_e = error.data) === null || _e === void 0 ? void 0 : _e.message) === null || _f === void 0 ? void 0 : _f.includes("Couldn't find that app.")) {
                this.error(new Error(`Couldn't find that app  ${appName}`));
            }
            this.error(e);
        }
    }
}
exports.default = LogDrainRemove;
LogDrainRemove.summary = messages.getMessage('summary');
LogDrainRemove.description = messages.getMessage('description');
LogDrainRemove.examples = messages.getMessages('examples');
LogDrainRemove.flags = {
    'target-compute': flags_1.FunctionsFlagBuilder.environment({
        exclusive: ['environment'],
    }),
    environment: flags_1.FunctionsFlagBuilder.environment({
        char: 'e',
        exclusive: ['target-compute'],
        hidden: true,
    }),
    'drain-url': core_1.Flags.string({
        exclusive: ['url'],
        char: 'l',
        description: messages.getMessage('flags.drain-url.summary'),
    }),
    url: core_1.Flags.string({
        exclusive: ['drain-url'],
        char: 'u',
        description: messages.getMessage('flags.drain-url.summary'),
        hidden: true,
    }),
};
