// @sf/env/lib/shared/sandboxProgress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxProgress = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const os = require("os");
const core_1 = require("@oclif/core");
const timeUtils_1 = require("../utils/timeUtils");
const stagedProgress_1 = require("./stagedProgress");
const columns = {
    key: { header: 'Field' },
    value: { header: 'Value' },
};
class SandboxProgress extends stagedProgress_1.StagedProgress {
    constructor(stageNames = ['Pending', 'Processing', 'Activating', 'Authenticating']) {
        super(stageNames);
    }
    getLogSandboxProcessResult(result) {
        const { sandboxProcessObj } = result;
        const sandboxReadyForUse = `Sandbox ${sandboxProcessObj.SandboxName}(${sandboxProcessObj.Id}) is ready for use.`;
        return sandboxReadyForUse;
    }
    getTableDataFromProcessObj(authUserName, sandboxProcessObj) {
        return [
            { key: 'Id', value: sandboxProcessObj.Id },
            { key: 'SandboxName', value: sandboxProcessObj.SandboxName },
            { key: 'Status', value: sandboxProcessObj.Status },
            { key: 'CopyProgress', value: `${sandboxProcessObj.CopyProgress}%` },
            { key: 'Description', value: sandboxProcessObj.Description },
            { key: 'LicenseType', value: sandboxProcessObj.LicenseType },
            { key: 'SandboxInfoId', value: sandboxProcessObj.SandboxInfoId },
            { key: 'SourceId', value: sandboxProcessObj.SourceId },
            { key: 'SandboxOrg', value: sandboxProcessObj.SandboxOrganization },
            { key: 'Created Date', value: sandboxProcessObj.CreatedDate },
            { key: 'ApexClassId', value: sandboxProcessObj.ApexClassId },
            { key: 'Authorized Sandbox Username', value: authUserName },
        ].filter((v) => !!v.value);
    }
    getSandboxProgress(event) {
        const statusUpdate = event;
        const waitingOnAuth = statusUpdate.waitingOnAuth ?? false;
        const { sandboxProcessObj } = event;
        const waitTimeInSec = statusUpdate.remainingWait ?? 0;
        const sandboxIdentifierMsg = `${sandboxProcessObj.SandboxName}(${sandboxProcessObj.Id})`;
        return {
            id: sandboxIdentifierMsg,
            status: waitingOnAuth || sandboxProcessObj.Status === 'Completed' ? 'Authenticating' : sandboxProcessObj.Status,
            percentComplete: sandboxProcessObj.CopyProgress,
            remainingWaitTime: waitTimeInSec,
            remainingWaitTimeHuman: waitTimeInSec === 0 ? '' : `${(0, timeUtils_1.getClockForSeconds)(waitTimeInSec)} until timeout.`,
        };
    }
    getSandboxTableAsText(sandboxUsername, sandboxProgress) {
        if (!sandboxProgress) {
            return [];
        }
        const tableRows = [];
        core_1.CliUx.ux.table(this.getTableDataFromProcessObj(sandboxUsername, sandboxProgress), columns, {
            printLine: (s) => {
                tableRows.push(s);
            },
        });
        return tableRows;
    }
    formatProgressStatus(withClock = true) {
        const table = this.getSandboxTableAsText(undefined, this.statusData.sandboxProcessObj).join(os.EOL);
        return [
            withClock
                ? `${(0, timeUtils_1.getClockForSeconds)(this.statusData.sandboxProgress.remainingWaitTime)} until timeout. ${this.statusData.sandboxProgress.percentComplete}%`
                : undefined,
            table,
            '---------------------',
            'Sandbox Create Stages',
            this.formatStages(),
        ]
            .filter((line) => line)
            .join(os.EOL);
    }
    mapCurrentStage(currentStage) {
        switch (currentStage) {
            case 'Pending Remote Creation':
                return 'Pending';
            case 'Remote Sandbox Created':
                return 'Pending';
            case 'Completed':
                return 'Authenticating';
            default:
                return currentStage;
        }
    }
}
exports.SandboxProgress = SandboxProgress;
