// @salesforce/plugin-packaging/lib/commands/force/package1/beta/version/list.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package1VersionListCommand = void 0;
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const packaging_1 = require("@salesforce/packaging");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package1_version_list');
class Package1VersionListCommand extends command_1.SfdxCommand {
    async run() {
        const result = await (0, packaging_1.package1VersionList)(this.org.getConnection(), this.flags.packageid);
        if (result.length) {
            this.ux.table(result, {
                MetadataPackageVersionId: { header: 'MetadataPackageVersionId' },
                MetadataPackageId: { header: 'MetadataPackageId' },
                Name: { header: 'Name' },
                Version: { header: 'Version' },
                ReleaseState: { header: 'ReleaseState' },
                BuildNumber: { header: 'BuildNumber' },
            });
        }
        else {
            this.ux.log('No Results Found');
        }
        return result;
    }
}
exports.Package1VersionListCommand = Package1VersionListCommand;
Package1VersionListCommand.description = messages.getMessage('cliDescription');
Package1VersionListCommand.requiresUsername = true;
Package1VersionListCommand.flagsConfig = {
    packageid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('packageId'),
        longDescription: messages.getMessage('packageIdLong'),
        validate: (id) => {
            if (id.startsWith('033') && [18, 15].includes(id.length)) {
                return true;
            }
            else {
                throw new core_1.SfError(messages.getMessage('packageIdInvalid'));
            }
        },
    }),
};
