// @salesforce/plugin-packaging/lib/commands/force/package/beta/version/promote.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionPromoteCommand = void 0;
const os = require("os");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const packaging_1 = require("@salesforce/packaging");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-packaging', 'package_version_promote');
class PackageVersionPromoteCommand extends command_1.SfdxCommand {
    async run() {
        var _a;
        const conn = this.hubOrg.getConnection();
        const packageId = (_a = (0, packaging_1.getPackageIdFromAlias)(this.flags.package, this.project)) !== null && _a !== void 0 ? _a : this.flags.package;
        // ID can be 04t or 05i at this point
        (0, packaging_1.validateId)([packaging_1.BY_LABEL.SUBSCRIBER_PACKAGE_VERSION_ID, packaging_1.BY_LABEL.PACKAGE_VERSION_ID], packageId);
        if (!this.flags.json && !this.flags.noprompt) {
            // Warn when a Managed package has removed metadata
            if (await (0, packaging_1.getHasMetadataRemoved)(packageId, conn)) {
                this.ux.warn(messages.getMessage('hasMetadataRemovedWarning'));
            }
            // Prompt for confirmation
            if (!(await this.ux.confirm(messages.getMessage('packageVersionPromoteConfirm', [this.flags.package])))) {
                return;
            }
        }
        const pkg = new packaging_1.PackageVersion({ connection: conn, project: this.project });
        let result;
        try {
            result = await pkg.promote(packageId);
        }
        catch (e) {
            const err = core_1.SfError.wrap(e);
            if (err.name === 'DUPLICATE_VALUE' && err.message.includes('previously released')) {
                err.message = messages.getMessage('previouslyReleasedMessage');
                err.actions = [messages.getMessage('previouslyReleasedAction')];
            }
            throw err;
        }
        result.id = packageId;
        this.ux.log(messages.getMessage('humanSuccess', [result.id]));
        return result;
    }
}
exports.PackageVersionPromoteCommand = PackageVersionPromoteCommand;
PackageVersionPromoteCommand.description = messages.getMessage('cliDescription');
PackageVersionPromoteCommand.examples = messages.getMessage('examples').split(os.EOL);
PackageVersionPromoteCommand.requiresDevhubUsername = true;
PackageVersionPromoteCommand.requiresProject = true;
PackageVersionPromoteCommand.flagsConfig = {
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package'),
        longDescription: messages.getMessage('packageLong'),
        required: true,
    }),
    noprompt: command_1.flags.boolean({
        char: 'n',
        description: messages.getMessage('setasreleasedForce'),
        longDescription: messages.getMessage('setasreleasedForceLong'),
    }),
};
