// @sf/sobject/lib/shared/convert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseXml = exports.convertJsonToXml = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const jsToXml = require("js2xmlparser");
const fast_xml_parser_1 = require("fast-xml-parser");
const standardOptions = {
    declaration: {
        include: true,
        encoding: 'UTF-8',
        version: '1.0',
    },
    format: {
        doubleQuotes: true,
    },
};
const convertJsonToXml = ({ json, type, options = standardOptions }) => jsToXml.parse(type, fixExistingDollarSign(json), options);
exports.convertJsonToXml = convertJsonToXml;
const fixExistingDollarSign = (existing) => {
    const existingCopy = { ...existing };
    if ('$' in existingCopy) {
        const temp = existingCopy.$;
        delete existingCopy.$;
        existingCopy['@'] = temp;
    }
    return existingCopy;
};
const parseXml = (fileString, mainNodeName) => {
    const parser = new fast_xml_parser_1.XMLParser();
    return parser.parse(fileString)[mainNodeName];
};
exports.parseXml = parseXml;
